codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/protocol/ServerProtocolTestGenerator.kt

```diff
@@ -298,22 +296,77 @@ class ServerProtocolTestGenerator(
         rustTemplate(
             """
-            #{AssertEq}(
-                http::StatusCode::from_u16(${testCase.code}).expect("invalid expected HTTP status code"),
-                http_response.status()
-            );
+            use #{AxumCore}::extract::FromRequest;
+            let mut http_request = #{AxumCore}::extract::RequestParts::new(http_request);
+            let rejection = super::$operationName::from_request(&mut http_request).await.expect_err("request was accepted but we expected it to be rejected");
+            use #{AxumCore}::response::IntoResponse;
+            let http_response = rejection.into_response();
+            """,
+            *codegenScope,
+        )
+        checkResponse(this, testCase.response)
+    }
+
+    private fun RustWriter.renderHttpRequest(
+        uri: String,
+        headers: Map<String, String>,
+        body: String?,
+        queryParams: List<String>,
+        host: String?
+    ) {
+        rustTemplate(
+            """
+            ##[allow(unused_mut)]
+            let mut http_request = http::Request::builder()
+                .uri("$uri")
             """,
             *codegenScope
         )
-        checkHeaders(this, "&http_response.headers()", testCase.headers)
-        checkForbidHeaders(this, "&http_response.headers()", testCase.forbidHeaders)
-        checkRequiredHeaders(this, "&http_response.headers()", testCase.requireHeaders)
-        checkHttpResponseExtensions(this)
-        // "If no request body is defined, then no assertions are made about the body of the message."
-        if (testCase.body.isPresent) {
-            checkBody(this, testCase.body.get(), testCase.bodyMediaType.orNull())
+        for (header in headers) {
+            rust(".header(${header.key.dq()}, ${header.value.dq()})")
+        }
+        rustTemplate(
+            """
+            .body(${
+                if (body != null) {
+                    // The `replace` is necessary to fix the malformed request test `RestJsonInvalidJsonBody`.
+                    // https://github.com/awslabs/smithy/blob/887ae4f6d118e55937105583a07deb90d8fabe1c/smithy-aws-protocol-tests/model/restJson1/malformedRequests/malformed-request-body.smithy#L47
+                    //
+                    // Smithy is written in Java, which parses `\u000c` within a `String` as a single char given by the
+                    // corresponding Unicode code point. That is the "form feed" 0x0c character. When printing it,
+                    // it gets written as "\f", which is an invalid Rust escape sequence: https://static.rust-lang.org/doc/master/reference.html#literals
+                    // So we need to write the corresponding Rust Unicode escape sequence to make the program compile.
+                    "#{SmithyHttpServer}::body::Body::from(#{Bytes}::from_static(b${body.replace("\u000c", "\\u{000c}").dq()}))"
```

Comment:
Is there a better way to do this?

Reply:
The detailed explanation helps a lot. If it makes you feel better, I would have implemented it the same way :) Probably there is some crate out in the wild that can help with this, but since it is a single occurrence in our code (so far), I believe we are good for now.